src/Compilers/Core/Portable/Operations/CustomDataFlowAnalysis.cs

```diff
@@ -0,0 +1,402 @@
+    {
+        /// <summary>
+        /// Runs dataflow analysis for the given <paramref name="analyzer"/> on the given <paramref name="blocks"/>.
+        /// </summary>
+        /// <param name="blocks">Blocks on which to execute analysis.</param>
+        /// <param name="analyzer">Dataflow analyzer.</param>
+        /// <returns>Block analysis data for the last block.</returns>
+        internal static TBlockAnalysisData Run(ImmutableArray<TBasicBlock> blocks, TDataFlowAnalyzer analyzer, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var continueDispatchAfterFinally = PooledDictionary<ControlFlowRegion, bool>.GetInstance();
+            var dispatchedExceptionsFromRegions = PooledHashSet<ControlFlowRegion>.GetInstance();
+            int firstBlockOrdinal = 0;
+            int lastBlockOrdinal = blocks.Length - 1;
+
+            var unreachableBlocksToVisit = ArrayBuilder<TBasicBlock>.GetInstance();
+            if (analyzer.AnalyzeUnreachableBlocks)
+            {
+                for (int i = firstBlockOrdinal; i <= lastBlockOrdinal; i++)
+                {
+                    if (!blocks[i].IsReachable)
+                    {
+                        unreachableBlocksToVisit.Add(blocks[i]);
+                    }
+                }
+            }
+
+            var initialAnalysisData = analyzer.GetCurrentAnalysisData(blocks[0]);
+            if (initialAnalysisData == default)
+            {
+                initialAnalysisData = analyzer.GetEmptyAnalysisData();
+            }
+
+            var result = RunCore(blocks, analyzer, firstBlockOrdinal, lastBlockOrdinal,
+                                 initialAnalysisData,
+                                 unreachableBlocksToVisit,
+                                 outOfRangeBlocksToVisit: null,
+                                 continueDispatchAfterFinally,
+                                 dispatchedExceptionsFromRegions,
+                                 cancellationToken);
+            Debug.Assert(unreachableBlocksToVisit.Count == 0);
+            unreachableBlocksToVisit.Free();
+            continueDispatchAfterFinally.Free();
+            dispatchedExceptionsFromRegions.Free();
+            return result;
+        }
+
+        private static TBlockAnalysisData RunCore(
+            ImmutableArray<TBasicBlock> blocks,
+            TDataFlowAnalyzer analyzer,
+            int firstBlockOrdinal,
+            int lastBlockOrdinal,
+            TBlockAnalysisData initialAnalysisData,
+            ArrayBuilder<TBasicBlock> unreachableBlocksToVisit,
+            SortedSet<int> outOfRangeBlocksToVisit,
+            PooledDictionary<ControlFlowRegion, bool> continueDispatchAfterFinally,
+            PooledHashSet<ControlFlowRegion> dispatchedExceptionsFromRegions,
+            CancellationToken cancellationToken)
+        {
+            var toVisit = new SortedSet<int>();
```

Comment:
`toVisit` was changed from a basic block stack to ordered set of basic blocks as it improved the performance of dataflow analysis. Otherwise, we end up analyzing blocks further down in the graph before the earlier blocks, leading to multiple iterations for fix point computation.